syntax = "proto3";

package foreverbull.backtest;

option go_package = "github.com/lhjnilsson/foreverbull/pkg/backtest/pb";

import "foreverbull/backtest/backtest.proto";
import "foreverbull/finance/finance.proto";
import "foreverbull/backtest/execution.proto";
import "foreverbull/backtest/ingestion.proto";

message GetIngestionRequest {
}

message GetIngestionResponse {
    Ingestion ingestion = 1;
}

message DownloadIngestionRequest {
    string bucket = 1;
    string object = 2;
}

message DownloadIngestionResponse {
    Ingestion ingestion = 1;
}

message IngestRequest {
    Ingestion ingestion = 1;
    optional string bucket = 2;
    optional string object = 3;
}

message IngestResponse {
}

message NewSessionRequest {
    string id = 1;
}

message NewSessionResponse {
    int64 port = 1;
}

message RunBacktestRequest {
    Backtest backtest = 1;
}

message RunBacktestResponse {
    Backtest backtest = 1;
}

message GetCurrentPeriodRequest {
}

message GetCurrentPeriodResponse {
    bool is_running = 1;
    optional foreverbull.finance.Portfolio portfolio = 2;
}

message PlaceOrdersAndContinueRequest {
    repeated foreverbull.finance.Order orders = 1;
}

message PlaceOrdersAndContinueResponse {
}

message GetResultRequest {
    string execution = 1;
    bool upload = 2;
}

message GetResultResponse {
    repeated foreverbull.backtest.Period periods = 1;
}


service Engine {
    rpc GetIngestion(GetIngestionRequest) returns (GetIngestionResponse) {}
    rpc DownloadIngestion(DownloadIngestionRequest) returns (DownloadIngestionResponse) {}
    rpc Ingest(IngestRequest) returns (IngestResponse) {}
    rpc NewSession(NewSessionRequest) returns (NewSessionResponse) {}
}

service EngineSession {
    rpc RunBacktest(RunBacktestRequest) returns (RunBacktestResponse) {}
    rpc GetCurrentPeriod(GetCurrentPeriodRequest) returns (GetCurrentPeriodResponse) {}
    rpc PlaceOrdersAndContinue(PlaceOrdersAndContinueRequest) returns (PlaceOrdersAndContinueResponse) {}
    rpc GetResult(GetResultRequest) returns (GetResultResponse) {}
}
